src/v/cluster/rm_stm.cc

```diff
@@ -1941,8 +1941,8 @@ model::offset rm_stm::last_stable_offset() {
         // transactions.
         return std::min(first_tx_start, next_to_apply);
     }
-    // no inflight transactions.
-    return model::next_offset(last_visible_index);
+    // no inflight transactions until last_applied (incl.)
+    return next_to_apply;
```

Comment:
The fix makes sense but I think this regresses the performance of non tx ..acks 0,1 (vaguely remember running into this when making this fix long time back).. LSO is blocked until every change is applied here... we should probably check if there are any tx batches to be applied in [last_applied, last_visible] range. If there are none, return next(last_visible), else return max(next_to_apply, next(tx_to_be_applied_begin)].. or something like that..

Reply:
I've added a perf fix in the next commit, thanks!